src/components/transform-object-button.js

```diff
@@ -31,6 +27,131 @@ function qAlmostEquals(a, b) {
   );
 }
 
+const calculatePlaneMatrix = (function() {
+  const planeMatrix = new THREE.Matrix4();
+  const planeUp = new THREE.Vector3();
+  const planeForward = new THREE.Vector3();
+  const planeRight = new THREE.Vector3();
+  const planePosition = new THREE.Vector3();
+  const camPosition = new THREE.Vector3();
+
+  return function calculatePlaneMatrix(camera, button) {
+    camera.updateMatrices();
+    camPosition.setFromMatrixPosition(camera.matrixWorld);
+    button.updateMatrices();
+    planePosition.setFromMatrixPosition(button.matrixWorld);
+    planeForward.subVectors(planePosition, camPosition);
+    planeForward.y = 0;
+    planeForward.normalize();
+    planeUp.set(0, 1, 0);
+    planeRight.crossVectors(planeForward, planeUp);
+    planeMatrix.makeBasis(planeRight, planeUp, planeForward.multiplyScalar(-1));
+    planeMatrix[12] = planePosition[0];
+    planeMatrix[13] = planePosition[1];
+    planeMatrix[14] = planePosition[2];
+    return planeMatrix;
+  };
+})();
+
+AFRAME.registerComponent("scale-button", {
+  init() {
+    this.isScaling = false;
+    this.plane = new THREE.Mesh(
+      new THREE.PlaneBufferGeometry(100000, 100000, 2, 2),
+      new THREE.MeshBasicMaterial({
+        visible: true,
+        wireframe: true,
+        side: THREE.DoubleSide,
+        transparent: true,
+        opacity: 0.3
+      })
+    );
+    this.planeRotation = new THREE.Matrix4();
+    this.planeUp = new THREE.Vector3();
+    this.planeRight = new THREE.Vector3();
+    this.intersections = [];
+    this.initialIntersectionPoint = new THREE.Vector3();
+    this.intersectionPoint = new THREE.Vector3();
+    this.initialObjectScale = new THREE.Vector3();
+    this.desiredObjectScale = new THREE.Vector3();
+    this.deltaScale = new THREE.Vector3();
+    this.objectMatrix = new THREE.Matrix4();
+    this.dragVector = new THREE.Vector3();
+    this.currentObjectScale = new THREE.Vector3();
+    NAF.utils.getNetworkedEntity(this.el).then(networkedEl => {
+      this.objectToScale = networkedEl.object3D;
+    });
+    this.startScaling = e => {
+      if (this.isScaling || !this.objectToScale) {
+        return;
+      }
+      if (!this.didGetObjectReferences) {
+        this.didGetObjectReferences = true;
+        this.leftEventer = document.getElementById("left-cursor").object3D;
+        this.leftRaycaster = document.getElementById("left-cursor-controller").components[
+          "cursor-controller"
+        ].raycaster;
+        this.rightRaycaster = document.getElementById("right-cursor-controller").components[
+          "cursor-controller"
+        ].raycaster;
+        this.viewingCamera = document.getElementById("viewing-camera").object3D;
+      }
+      setMatrixWorld(this.plane, calculatePlaneMatrix(this.viewingCamera, this.el.object3D));
+      this.planeRotation.extractRotation(this.plane.matrixWorld);
```

Comment:
this seems like it can be extracted into another function

Reply:
(and used in tick)

Reply:
sorry meant line 104+

Reply:
Good point! :+1: